File: fun-readersAndWriters/src/main/java/com/practice/useakka/Reader.java
package com.practice.useakka;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import lombok.RequiredArgsConstructor;

public class Reader {

    public interface Command {
    }

    @RequiredArgsConstructor
    public static class ReadMessage implements Command {
        public final ActorRef<Mail.Command> readFrom;
    }

    @RequiredArgsConstructor
    public static class WrappedMailCommand implements Command {
        public final Mail.Response command;
    }

    public static Behavior<Command> create(String name) {
        return Behaviors.setup(context -> new Reader(context, name).reader());
    }

    private final ActorContext<Reader.Command> context;
    private final String name;
    private final ActorRef<Mail.Response> messageAdapter;

    public Reader(ActorContext<Command> context, String name) {
        this.context = context;
        messageAdapter = context.messageAdapter(Mail.Response.class, WrappedMailCommand::new);
        this.name = name;
    }

    public Behavior<Command> reader() {
        return Behaviors.receive(Command.class)
                .onMessage(WrappedMailCommand.class, this::onWrappedMessageRespond)
                .onMessage(ReadMessage.class, this::onReadMessage)
                .build();
    }

    private Behavior<Command> onReadMessage(ReadMessage m) {
        m.readFrom.tell(new Mail.GetMessage(messageAdapter));
        return Behaviors.same();
    }

    private Behavior<Command> onWrappedMessageRespond(WrappedMailCommand m) {
        final Mail.MessageRespond messageRespond = m.command instanceof Mail.MessageRespond ? ((Mail.MessageRespond) m.command) : null;
        if (messageRespond != null) {
            context.getLog().info("{}: {}", name, messageRespond.message);
        }
        return Behaviors.same();
    }
}

File: scala-readersAndWriters/src/main/scala/com/practice/useakka/Reader.scala
package com.practice.useakka

import akka.actor.typed.{ActorRef, Behavior}
import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}

object Reader {
  sealed trait Command
  object ReadMessage extends Command
  private final case class WrappedMailResponse(response: Mail.Response) extends Command

  def apply(mail: ActorRef[Mail.Command]): Behavior[Command] = Behaviors.setup(context => new Reader(context, mail))
}

class Reader(context: ActorContext[Reader.Command], mail: ActorRef[Mail.Command])
  extends AbstractBehavior[Reader.Command](context) {

  import Reader._

  private val mailResponseAdapter: ActorRef[Mail.Response] =
    context.messageAdapter(response => WrappedMailResponse(response))

  override def onMessage(msg: Command): Behavior[Command] = msg match {
    case ReadMessage =>
      mail.tell(Mail.PopMessage(mailResponseAdapter))
      this
    case WrappedMailResponse(response) =>
      response match {
        case Mail.MessageRespond(message) =>
          context.log.info("Read message: {}", message)
          this
      }
  }
}
